### ISSUE TITLE
Correct approach to sync full tree of zfs

### ISSUE THREAD

**Issue by Peter2121** [bolthole/zrep#210]
I would like to create a script to sync the full tree of all zfs filesystems of a newly installed FreeBSD.

The list of the filesystems is as follows:

```
zroot
zroot/ROOT
zroot/ROOT/default
zroot/home
zroot/home/peter
zroot/tmp
zroot/usr
zroot/usr/ports
zroot/usr/src
zroot/var
zroot/var/audit
zroot/var/crash
zroot/var/log
zroot/var/mail
zroot/var/tmp
```

Another filesystems could be created during install or after it (for example, `zroot/home/<user>` or `zroot/ROOT/default/<update>`). The existing zrep configuration should continue to work in such case, only new filesystems need to be added to the sync list.

As I understand, the recursive replication is not reliable, the filesystems should be synced one-by-one. As I mentioned in #166, zrep data is propagated to the child filesystems, so the only approach would be ZREPTAG defined separately for EVERY filesystem to sync. It looks a bit hard to manage and snapshot deletion needs to be debugged/validated, so I would like to ask if any other approach looks possible before starting developing this solution.

**Comment by ppbrown** [comment]
Errrr...
some of those filesystems are a waste of time and space to back up. Example:

var crash
var tmp
tmp
usr/ports


Usually, best practices are to have ONE filesystem with stuff you actually
care about, and then keep a config of settings for the system.

This gives you the advantage that when and if you have a full system
meltdown, you can then rebuild a newer OS installation cleanly... then just
restore /home.

**Comment by ppbrown** [comment]
PS:
I would imagine that doing a "zrep -R" type backup of /home should work
okay.
Dont have a separate one for /home/peter, and everything should be good I
think?

**Comment by Peter2121** [comment]
> some of those filesystems are a waste of time and space to back up

Sure, I don't plan to sync (for example) `zroot/tmp`

> This gives you the advantage that when and if you have a full system meltdown, you can then rebuild a newer OS installation cleanly

To accelerate recovery, it would be nice to be able to back-sync most of filesystems. Like this, one can just boot a live-system on a VM with minimal system installed (for example, VM created from a template) and sync remote data to them. A simple script permits to restore the whole VM very fast (only the network is the bottleneck).

> Dont have a separate one for /home/peter

Since 14.1 (?) there are one ZFS dataset per user and one dataset for `/home`

Anyway, I can imagine another root dataset, for example `zdata/NFSROOT` with many children organized in many levels like `zdata/NFSROOT/db/mariadb/prod` `zdata/NFSROOT/documents/sales` and so on...

**Comment by ppbrown** [comment]
sorry for the ambiguity: I just meant "dont have a separate zrep call
for /home/peter".

I think that backing up homedirs like that is exactly what I
originally added the -R feature for, many years ago :)

**Comment by Peter2121** [comment]
As you've mentioned in #165 

> yeah... Im not happy with the way things work.... but zrep depends on ZFS. and recursive syncs with ZFS are.... tricky.

> I always recommend that people do syncs individually, not recursive, whenever possible.

...I don't try to use -R following your recommendations ;)

**Comment by ppbrown** [comment]
fair enough... :) but in this specific case, I would recommend you try it
out just for /home

Unless you really want to do the "back up ALL filesystems in one go" thing.
Even though it is wasteful.
In which case, you would only want to do the top level.
But.. this would be experimental, and you should do lots of use-case
testing,  and not rely on it 100%
